pkg/geth/node/node.go

```diff
@@ -5,45 +5,82 @@ import (
 
 	"strconv"
 
+	"regexp"
+
 	"github.com/ethereum/go-ethereum/core/types"
 	"github.com/ethereum/go-ethereum/p2p"
 	"github.com/ethereum/go-ethereum/rpc"
 	"github.com/vulcanize/vulcanizedb/pkg/core"
 )
 
-func Info(client *rpc.Client) core.Node {
+type Getter interface {
+	NetworkId() float64
+	GenesisBlock() string
+	GethNodeInfo() (string, string)
+	ParityNodeInfo() (string, string)
+}
+
+type NodeFactory struct {
+	Client *rpc.Client
+}
+
+func MakeNode(getter Getter) core.Node {
 	node := core.Node{}
-	node.NetworkID = NetworkID(client)
-	node.GenesisBlock = GenesisBlock(client)
-	node.ID, node.ClientName = IDClientName(client)
+	node.NetworkID = getter.NetworkId()
+	node.GenesisBlock = getter.GenesisBlock()
+	node.ID, node.ClientName = getter.GethNodeInfo()
+	if notGethNode(node) {
+		node.ID, node.ClientName = getter.ParityNodeInfo()
+	}
 	return node
 }
 
-func IDClientName(client *rpc.Client) (string, string) {
+func notGethNode(node core.Node) bool {
+	return node.ID == ""
+}
+
+func (nf *NodeFactory) GethNodeInfo() (string, string) {
 	var info p2p.NodeInfo
-	modules, _ := client.SupportedModules()
+	modules, _ := nf.Client.SupportedModules()
 	if _, ok := modules["admin"]; ok {
-		client.CallContext(context.Background(), &info, "admin_nodeInfo")
+		nf.Client.CallContext(context.Background(), &info, "admin_nodeInfo")
 		return info.ID, info.Name
 	}
 	return "", ""
 }
 
-func NetworkID(client *rpc.Client) float64 {
+func (nf *NodeFactory) NetworkId() float64 {
 	var version string
-	client.CallContext(context.Background(), &version, "net_version")
+	nf.Client.CallContext(context.Background(), &version, "net_version")
 	networkId, _ := strconv.ParseFloat(version, 64)
 	return networkId
 }
 
-func ProtocolVersion(client *rpc.Client) string {
-	var protocolVersion string
-	client.CallContext(context.Background(), &protocolVersion, "eth_protocolVersion")
-	return protocolVersion
+func (nf *NodeFactory) ParityNodeInfo() (string, string) {
+	client := nf.parityClient()
+	id := nf.parityId()
+	return id, client
+}
+
+func (nf *NodeFactory) parityClient() string {
+	var nodeInfo core.ParityNodeInfo
+	nf.Client.CallContext(context.Background(), &nodeInfo, "parity_versionInfo")
+	return nodeInfo.String()
+}
+
+func (nf *NodeFactory) parityId() string {
+	var enodeId = regexp.MustCompile(`^enode://(.+)@.+$`)
+	var enodeURL string
+	nf.Client.CallContext(context.Background(), &enodeURL, "parity_enode")
+	enode := enodeId.FindStringSubmatch(enodeURL)
+	if len(enode) < 2 {
+		return ""
+	}
+	return enode[1]
 }
 
-func GenesisBlock(client *rpc.Client) string {
+func (nf *NodeFactory) GenesisBlock() string {
 	var header *types.Header
-	client.CallContext(context.Background(), &header, "eth_getBlockByNumber", "0x0", false)
+	nf.Client.CallContext(context.Background(), &header, "eth_getBlockByNumber", "0x0", false)
```

Comment:
Take it or leave it, but I'd consider wrapping the address and boolean val in vars that reveal what they mean when passed to `CallContext`

Reply:
makes sense. updated.